src/router/smcroute/man/smcroutectl.8
.\"  -*- nroff -*-
.Dd November 28, 2021
.Dt SMCROUTECTL 8 SMM
.Os
.Sh NAME
.Nm smcroutectl
.Nd Control and status tool for
.Xr smcrouted 8
.Sh SYNOPSIS
.Nm smcroutectl
.Op Fl bdptv
.Op Fl i Ar NAME
.Op Fl u Ar FILE
.Op Ar COMMAND
.Pp
.Nm smcroutectl
.Ao help | flush | kill | reload | version Ac
.Nm smcroutectl
.Ao show Ac
.Op groups | routes
.Nm smcroutectl
.Ao add \ | \ \ rem Ac IIF Oo SOURCE Oc Ar GROUP[/LEN] OIF Op OIF ...
.Nm smcroutectl
.Ao join | leave Ac IIF Oo SOURCE Oc Ar GROUP[/LEN]
.Sh DESCRIPTION
.Nm
is the control tool for
.Xr smcrouted 8 .
It can be used to query status, debug, modify the kernel multicast
forwarding cache (MFC), manage group interface memberships, reload
.Nm smcroute.conf ,
and kill a running
.Nm smcrouted .
.Sh OPTIONS
The following
.Nm
options are available:
.Bl -tag -width Ds
.It Fl b
Batch mode, read commands from stdin.
.Bd -unfilled -offset indent
$ sudo smcroutectl -b <<-EOF
	join eth0 225.1.2.3
	add eth0 192.168.1.42 225.1.2.3 eth1 eth2
	rem eth1 225.3.4.5 eth3
	leave eth1 225.3.4.5
	EOF
.Ed
.It Fl d
Enable detailed output in show commands.
.It Fl i Ar NAME
Connect to an
.Nm smcrouted
instance that runs with another identity,
.Ar NAME .
.Pp
This option is required for both
.Nm smcrouted
and
.Nm smcroutectl
when running multiple
.Nm smcrouted
instances, e.g., when using multiple routing tables, on Linux.
.It Fl p
Use plain table headings in
.Cm show
command output.  No ANSI control characters are used, not even for
probing screen width.
.It Fl t
Skip table headings entirely in
.Cm show
command output.
.It Fl u Ar FILE
UNIX domain socket path, used for the IPC between
.Nm smcrouted
and
.Nm .
Use this to override the default socket path, otherwise derived from the
identity,
.Fl i Ar NAME .
This option can be useful when overriding the identity is not
sufficient, e.g. for testing.  The default depends on how
.Nm
is configured at build time, see
.Sx FILES .
.El
.Sh OPERATION
The
.Ar IIF
and
.Ar OIF
arguments in the below
.Nm smcroutectl
commands are the interface names, or interface wildcards of the form
.Ar eth+ ,
which matches
.Ar eth0 , eth10 ,
etc.  Wildcards are available for both inbound and outbound interfaces.
.Pp
A multicast route is defined by an input interface
.Ar IIF ,
the sender's unicast IP address
.Ar SOURCE ,
which is optional, the multicast group
.Ar GROUP
and a list of, at least one, output interface
.Ar OIF [OIF ...] .
.Pp
Please refer to
.Xr smcrouted 8 
for more details on the operation and how ASM/SSM multicast works.
.Sh COMMANDS
Commands can be abbreviated to the minimum unambiguous prefix; for
example,
.Cm s g
for
.Cm show groups .
The following commands are available:
.Bl -tag -width Ds
.It Nm add Ar IIF [SOURCE[/LEN]] GROUP[/LEN] OIF [OIF ...]
Add a new multicast route the the kernel MFC, or modify the outbound
interfaces (OIF) an existing route.
.Pp
The arguments are, in order:
.Ar IIF
the inbound interface,
.Ar SOURCE
originating IP address (may need to be reachable in the unicast routing
table to be allowed by the kernel reverse-path check),
.Ar GROUP
the multicast group address, and
.Ar OIF Oo Ar OIF ... Oc
the outbound network interface(s).
.Pp
The interfaces provided as
.Ar IIF
and
.Ar OIF
can be any multicast capable network interface as listed by
.Ql Cm ifconfig
or
.Ql Cm ip link list ,
including tunnel interfaces and loopback.  Provided
.Nm smcrouted
has "enumerated" them.  See
.Xr smcrouted 8 ,
in particular the command line option
.Fl N ,
and the
.Xr smcroute.conf 5
.Ql Cm phyint
directive.
.Pp
To add a (*,G) route, either omit the
.Ar SOURCE
argument completely, or set it to
.Ar 0.0.0.0
for IPv4, and if you want to specify a range of groups, use
the
.Ql GROUP/LEN
modifier, e.g.
.Ql 225.0.0.0/24 .
.It Nm remove Ar IIF [SOURCE[/LEN]] GROUP[/LEN] [OIF [OIF ...]]
Remove or modify the outbound interfaces of a multicast route in the
kernel MFC.
.Pp
When no
.Ar OIF
argument is given, this command removes the entire route.  With
one or more
.Ar OIF
arguments, each outbound interface listed is removed.  Skipping
any unmatched or invalid interface names.  When no more outbound
interfaces exist, the route will have been transformed into a
"stop filter".  To remove the route entirely, the command must
be given with no
.Ar OIF
arguments.
.It Nm flush
Flush dynamic (*,G) multicast routes now.  Similar to how
.Fl c Ar SEC
works in
.Nm smcrouted ,
this command initiates an immediate flush of all dynamically installed
(*,G) multicast routes.  Useful when a topology change has been detected
and need to be propagated to
.Nm smcrouted.
.It Nm join Ar IIF [SOURCE[/LEN]] GROUP[/LEN]
Join a multicast group, with an optional prefix length, on the given
(inbound) interface.  The source address is optional, but if given a
source specific (SSM) join is performed.  Note, joining groups is only
ever necessary on the inbound interface, never on the outbound.  Unless,
two-way routing the same group.
.Pp
Note, as mentioned in
.Xr smcrouted 8 ,
joining a group to open up traffic in layer-2 network switches is only a
workaround to direct multicast towards SMCRoute.  When routing lots of
traffic it is advised to avoid this mechanism.  Instead, use multicast
router ports, or similar settings on the switches, or if they support
multicast router discovery (MRDISC), see RFC4286.
.It Nm leave Ar IIF [SOURCE[/LEN]] GROUP[/LEN]
Leave a multicast group, with optional prefix length, on a given
(inbound) interface.  As with the join command, above, the source
address is optional, but if the group was subscribed to with source it
must be unsubscribed with source as well.
.It Nm help [cmd]
Print a usage information message.
.It Nm kill
Tell a running
.Nm smcrouted
to exit gracefully, same as
.Ar SIGTERM .
.It Nm reload
Tell
.Nm smcrouted
to reload its configuration and activate the changes.  Same as
.Ar SIGHUP .
Note, any routes or groups added or removed with
.Nm smcroutectl
will be lost.  Only the configuration set in the file
.Pa smcroute.conf
is activated.
.It Nm show [groups|routes]
Show joined multicast groups or multicast routes, defaults to show
routes.  Can be combined with the
.Fl d
option to get details for each multicast route.
.It Nm version
Show program version and support information.
.El
.Sh SEE ALSO
.Xr smcrouted 8 ,
.Xr smcroute.conf 5
.Sh AUTHORS
.An -nosplit
SMCRoute was originally created by
.An Carsten Schill Aq Mt carsten@cschill.de .
Initial IPv6 support by
.An Todd Hayton Aq Mt todd.hayton@gmail.com .
Initial FreeBSD support by
.An Micha Lenk Aq Mt micha@debian.org .
.Pp
SMCRoute is currently maintained by
.An Joachim Wiberg Aq Mt troglobit@gmail.com ,
and
.An Micha Lenk Aq Mt micha@debian.org
at
.Lk https://github.com/troglobit/smcroute "GitHub" .


src/router/smcroute/src/smcroutectl.c
/* Client for smcrouted, not needed if only using smcroute.conf
 *
 * Copyright (C) 2011-2021  Joachim Wiberg <troglobit@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
 */

#include "config.h"

#include <err.h>
#include <errno.h>
#include <poll.h>
#include <stdio.h>
#include <stddef.h>
#include <string.h>
#include <sysexits.h>
#ifdef HAVE_TERMIOS_H
# include <termios.h>
#endif
#include <unistd.h>
#include <netinet/in.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>

#include "msg.h"
#include "util.h"

static const char version_info[] = PACKAGE_NAME " v" PACKAGE_VERSION;

static char *ident = PACKAGE;
static char *sock_file = NULL;
static char *prognm = NULL;
static int   heading = 1;
static int   detail = 0;
static int   plain = 0;
static int   help = 0;

struct arg {
	char *name;
	int   min_args;		/* 0: command takes no arguments */
	int   val;
	char *arg;
	char *help;
	char *example;		/* optional */
	int   has_detail;
} args[] = {
	{ NULL,      0, 'b', NULL,   "Batch mode, read commands from stdin", NULL, 0 },
	{ NULL,      0, 'd', NULL,   "Detailed output in show command", NULL, 0 },
	{ NULL,      1, 'i', "NAME", "Identity of routing daemon instance, default: " PACKAGE, "foo", 0 },
	{ NULL,      1, 'I', "NAME", NULL, NULL, 0 }, /* Alias, compat with older versions */
	{ NULL,      0, 'p', NULL,   "Use plain table headings, no ctrl chars", NULL, 0 },
	{ NULL,      0, 't', NULL,   "Skip table heading in show command", NULL, 0 },
	{ NULL,      1, 'u', "FILE", "UNIX domain socket for daemon, default: " RUNSTATEDIR "/" PACKAGE ".sock", "/tmp/foo.sock", 0 },
	{ "help",    0, 'h', NULL,   "Show help text", NULL, 0 },
	{ "version", 0, 'v', NULL,   "Show program version and support information", NULL, 0 },
	{ "flush" ,  0, 'F', NULL,   "Flush all dynamically installed (*,G) multicast routes", NULL, 0 },
	{ "kill",    0, 'k', NULL,   "Kill running daemon", NULL, 0 },
	{ "reload",  0, 'H', NULL,   "Reload .conf file, like SIGHUP", NULL, 0 },
	{ "restart", 0, 'H', NULL,   NULL, NULL, 0 }, /* Alias, compat with older versions */
	{ "show",    0, 's', NULL,   "Show status of routes, joined groups, interfaces, etc.", NULL, 1 },
	{ "add",     3, 'a', NULL,   "Add a multicast route",    "eth0 192.168.2.42 225.1.2.3 eth1 eth2", 0 },
	{ "remove",  2, 'r', NULL,   "Remove a multicast route", "eth0 192.168.2.42 225.1.2.3", 0 },
	{ "del",     2, 'r', NULL,   NULL, NULL, 0 }, /* Alias */
	{ "join",    2, 'j', NULL,   "Join multicast group on an interface", "eth0 225.1.2.3", 0 },
	{ "leave",   2, 'l', NULL,   "Leave joined multicast group",         "eth0 225.1.2.3", 0 },
	{ NULL, 0, 0, NULL, NULL, NULL, 0 }
};


/*
 * Build IPC message to send to the daemon using @cmd and @count
 * number of arguments from @argv.
 */
static struct ipc_msg *msg_create(uint16_t cmd, char *argv[], size_t count)
{
	struct ipc_msg *msg;
	size_t len = 0;
	size_t i, sz;
	char *ptr;

	for (i = 0; i < count; i++)
		len += strlen(argv[i]) + 1;

	sz = sizeof(struct ipc_msg) + len + 1;
	if (sz > MX_CMDPKT_SZ) {
		errno = EMSGSIZE;
		return NULL;
	}

	msg = calloc(1, sz);
	if (!msg)
		return NULL;

	msg->len   = sz;
	msg->cmd   = cmd;
	msg->count = count;

	ptr = (char *)msg->argv;
	for (i = 0; i < count; i++) {
		len = strlen(argv[i]) + 1;
		ptr = memcpy(ptr, argv[i], len) + len;
	}
	*ptr = '\0';	/* '\0' behind last string */

	return msg;
}

#define ESC "\033"
static int get_width(void)
{
	int ret = 79;
#ifdef HAVE_TERMIOS_H
	struct pollfd fd = { STDIN_FILENO, POLLIN, 0 };
	struct termios tc, saved;
	char buf[42];

	memset(buf, 0, sizeof(buf));
	tcgetattr(STDERR_FILENO, &tc);
	saved = tc;
	tc.c_cflag |= (CLOCAL | CREAD);
	tc.c_lflag &= ~(ICANON | ECHO | ECHOE | ISIG);
	tcsetattr(STDERR_FILENO, TCSANOW, &tc);
	fprintf(stderr, ESC "7" ESC "[r" ESC "[999;999H" ESC "[6n");

	if (poll(&fd, 1, 300) > 0) {
		int row, col;

		if (scanf(ESC "[%d;%dR", &row, &col) == 2)
			ret = col;
	}

	fprintf(stderr, ESC "8");
	tcsetattr(STDERR_FILENO, TCSANOW, &saved);
#endif
	return ret;
}

static void print(char *line, int indent)
{
	int type = 0;
	int i, len;

	chomp(line);

	/* Table headings, or repeat headers, end with a '=' */
	len = (int)strlen(line) - 1;
	if (len > 0) {
		if (line[len] == '_')
			type = 1;
		if (line[len] == '=')
			type = 2;

		if (type) {
			if (!heading)
				return;
			line[len] = 0;
		}
	}

	switch (type) {
	case 1:
		if (!plain) {
			fprintf(stdout, "\e[4m%*s\e[0m\n%s\n", get_width(), "", line);
			return;

		}

		len = len < 79 ? 79 : len;
		for (i = 0; i < len; i++)
			fputc('_', stdout);
		fprintf(stdout, "\n%*s%s\n", indent, "", line);
		break;

	case 2:
		if (!plain) {
			len = get_width() - len;
			fprintf(stdout, "\e[7m%s%*s\e[0m\n", line, len, "");
			return;
		}

		len = len < 79 ? 79 : len;
		for (i = 0; i < len; i++)
			fputc('=', stdout);
		fprintf(stdout, "\n%*s%s\n", indent, "", line);
		for (i = 0; i < len; i++)
			fputc('=', stdout);
		fputs("\n", stdout);
		break;

	default:
		puts(line);
		break;
	}
}

/*
 * Connects to the IPC socket of the server
 */
static int ipc_connect(char *path)
{
	struct sockaddr_un sa;
	socklen_t len;
	int sd;

	sd = socket(AF_UNIX, SOCK_STREAM, 0);
	if (sd < 0)
		return -1;

#ifdef HAVE_SOCKADDR_UN_SUN_LEN
	sa.sun_len = 0;	/* <- correct length is set by the OS */
#endif
	sa.sun_family = AF_UNIX;
	if (!path)
		snprintf(sa.sun_path, sizeof(sa.sun_path), "%s/%s.sock", RUNSTATEDIR, ident);
	else
		snprintf(sa.sun_path, sizeof(sa.sun_path), "%s", path);

	len = offsetof(struct sockaddr_un, sun_path) + strlen(sa.sun_path);
	if (connect(sd, (struct sockaddr *)&sa, len) < 0) {
		int err = errno;

		if (ENOENT == errno)
			warnx("Cannot find IPC socket %s", sa.sun_path);

		close(sd);
		errno = err;

		return -1;
	}

	return sd;
}

static int ipc_command(uint16_t cmd, char *argv[], size_t count)
{
	char buf[MX_CMDPKT_SZ + 1];
	struct ipc_msg *msg;
	int retries = 30;
	int result = 0;
	ssize_t total;
	ssize_t len;
	FILE *fp;
	int sd;

	msg = msg_create(cmd, argv, count);
	if (!msg) {
		warn("Failed constructing IPC command");
		return 1;
	}

	while ((sd = ipc_connect(sock_file)) < 0) {
		switch (errno) {
		case EACCES:
			warnx("Need root privileges to connect to daemon");
			break;

		case ECONNREFUSED:
			if (--retries) {
				usleep(100000);
				continue;
			}

			warnx("Daemon not running");
			break;

		case ENOENT:
			if (!sock_file) {
				warnx("Daemon may be running with another -i NAME");
				break;
			}
			/* fallthrough */

		default:
			warn("Failed connecting to daemon");
			break;
		}

		free(msg);
		return 1;
	}

	/* Send command */
	if (write(sd, msg, msg->len) != (ssize_t)msg->len) {
		warn("Communication with daemon failed");
		close(sd);
		free(msg);

		return 1;
	}

	fp = tempfile();
	if (!fp) {
		close(sd);
		free(msg);
		err(EX_OSERR, "Failed creating tempfile()");
	}

	total = 0;
	while ((len = read(sd, buf, sizeof(buf) - 1)) > 0) {
		total += len;
		buf[len] = 0;
		fwrite(buf, len, 1, fp);
	}
	rewind(fp);

	if (total > 1) {
		if (cmd == 'S' || cmd == 's') {
			while (fgets(buf, sizeof(buf), fp))
				print(buf, 0);
		} else {
			if (fgets(buf, sizeof(buf), fp))
				warnx("%s", buf);
			result = 1;
		}
	}

	fclose(fp);
	close(sd);
	free(msg);

	return result;
}

static int usage(int code)
{
	int i;

	printf("Usage:\n  %s [OPTIONS] CMD [ARGS]\n\n", prognm);

	printf("Options:\n");
	for (i = 0; args[i].val; i++) {
		if (!args[i].help)
			continue;

		if (args[i].name)
			continue;

		printf("  -%c %-10s %s\n", args[i].val, args[i].arg ? args[i].arg : "", args[i].help);
	}

	printf("\nCommands:\n");
	for (i = 0; args[i].val; i++) {
		if (!args[i].help)
			continue;

		if (!args[i].name)
			continue;

		printf("  %-7s %s  %s\n", args[i].name,
		       args[i].min_args ? "ARGS" : "    ", args[i].help);
	}

	printf("\nArguments:\n"
	       "         <---------- INBOUND ------------>  <- OUTBOUND ->\n"
	       "  add    IIF [SOURCE-IP[/LEN]] GROUP[/LEN]  OIF [OIF ... ]\n"
	       "  remove IIF [SOURCE-IP[/LEN]] GROUP[/LEN]\n"
	       "\n"
	       "  join   IIF [SOURCE-IP[/LEN]] GROUP[/LEN]\n"
	       "  leave  IIF [SOURCE-IP[/LEN]] GROUP[/LEN]\n"
	       "\n"
	       "  show   interfaces    Show configured multicast interfaces\n"
	       "  show   groups        Show joined multicast groups\n"
	       "  show   routes        Show (*,G) and (S,G) multicast routes, default\n"
	       "\n"
	       "Note:\n"
	       "  Inbound (IIF) and outbound (OIF) interfaces can be either an interface\n"
	       "  name or a wildcard.  E.g., \"eth+\" matches eth0, eth15, etc.\n"
	       "\n");

	return code;
}

static int version(void)
{
	puts(version_info);
	printf("\n"
	       "Bug report address: %s\n", PACKAGE_BUGREPORT);
#ifdef PACKAGE_URL
	printf("Project homepage:   %s\n", PACKAGE_URL);
#endif
	return 0;
}

static int parse(int pos, int argc, char *argv[])
{
	struct arg *cmd = NULL;
	int status = 0;
	int c;

	while (pos < argc && !cmd) {
		char *arg = argv[pos];
		int i;

		for (i = 0; args[i].val; i++) {
			char    *nm = args[i].name;
			size_t  len;

			if (!nm)
				continue;

			len = MIN(strlen(nm), strlen(arg));
			if (strncmp(arg, nm, len))
				continue;

			c = args[i].val;
			switch (c) {
			case 'h':
				help++;
				break;

			case 'v':
				return version();

			default:
				cmd = &args[i];
				if (help)
					goto help;
				if (argc - (pos + 1) < args[i].min_args) {
					warnx("Not enough arguments to command %s", nm);
					status = 1;
					goto help;
				}
				break;
			}

			break;	/* Next arg */
		}
		pos++;
	}

	if (help) {
		if (!cmd)
			return usage(0);
	help:
		while (!cmd->help)
			cmd--;
		printf("Help:\n"
		       "  %s\n\n"
		       "Example:\n"
		       "  %s %s %s\n\n", cmd->help,
		       prognm, cmd->name, cmd->example ? cmd->example : "");
		return status;
	}

	if (!cmd)
		return ipc_command(detail ? 'S' : 's', NULL, 0);

	c = cmd->val;
	if (detail && cmd->has_detail)
		c -= 0x20;

	return ipc_command(c, &argv[pos], argc - pos);
}

static int batch(void)
{
	char line[512];
	int rc = 0;

	while (fgets(line, sizeof(line), stdin)) {
		char *ptr, *token, *args[10];
		int num = 0;

		ptr = chomp(line);
		if (ptr[0] == '#')
			continue;

		while (num < 9 && (token = strsep(&ptr, " \t")))
			args[num++] = token;

		if (!num)
			continue;

		rc += parse(0, num, args);
	}

	return rc;
}

static char *progname(const char *arg0)
{
	char *nm;

	nm = strrchr(arg0, '/');
	if (nm)
		nm++;
	else
		nm = (char *)arg0;

	return nm;
}

int main(int argc, char *argv[])
{
	int batch_mode = 0;
	int c;

	prognm = progname(argv[0]);
	while ((c = getopt(argc, argv, "bdhI:i:ptu:v")) != EOF) {
		switch (c) {
		case 'b':
			batch_mode = 1;
			break;

		case 'd':
			detail++;
			break;

		case 'h':
			help++;
			break;

		case 'I':	/* compat with previous versions */
		case 'i':
			ident = optarg;
			break;

		case 'p':
			plain = 1;
			break;

		case 't':
			heading = 0;
			break;

		case 'u':
			sock_file = optarg;
			break;

		case 'v':
			return version();

		default:
			return usage(1);
		}
	}

	if (batch_mode)
		return batch();

	return parse(optind, argc, argv);
}

/**
 * Local Variables:
 *  indent-tabs-mode: t
 *  c-file-style: "linux"
 * End:
 */
